include/libtwofloat/arithmetics/double-word-arithmetic.hpp

```diff
@@ -233,5 +233,628 @@ inline two<T> div(const two<T> &x, const two<T> &y) {
+  */
+
+  T zeropointzero, onepointzero, pointfive, local_nan;
+  if constexpr (std::is_same_v<T, float>) {
+    onepointzero = 1.0f;
+    zeropointzero = 0.0f;
+    pointfive = 0.5f;
+    local_nan = fp32_nan;
+  } else if constexpr (std::is_same_v<T, double>) {
+    onepointzero = 1.0;
+    zeropointzero = 0.0;
+    pointfive = 0.5;
+    local_nan = fp64_nan;
+  } else {
+    std::cerr << "Other types not supported\n";
+  }
+
+  if (input.eval() == zeropointzero) {
+    T hi, lo;
+    qd::split(zeropointzero, hi, lo);
+    return two<T>{hi, lo};
+  }
+
+  // if (input.is_negative()) {
+  if (input.h < zeropointzero) {
+    std::cerr << "Negative argument.\n";
+    return two<T>{local_nan, local_nan};
+  }
+
+  T x = onepointzero / std::sqrt(input.h);
+  T ax = input.h * x;
+  two<T> temp = sub<doubleword::Mode::Accurate>(input, dd_real::sqr(ax));
+  return dd_real::add(ax, temp.h * x * pointfive);
+}
+
+// Reference: QD / dd_real.cpp
+template <typename T>
+static void sincos_taylor(const two<T> &input, two<T> &sin_a, two<T> &cos_a) {
+
+  T zeropointzero, onepointzero;
+  if constexpr (std::is_same_v<T, float>) {
+    zeropointzero = 0.0f;
+    onepointzero = 1.0f;
+  } else if constexpr (std::is_same_v<T, double>) {
+    zeropointzero = 0.0;
+    onepointzero = 1.0;
+  } else {
+    std::cerr << "Other types not supported\n";
+  }
+
+  // Reference: QD / dd_inline.h
+  // Assignment is performed according to operator== 
+  if (input.eval() == zeropointzero) {
+    sin_a.h = zeropointzero;
+    sin_a.l = zeropointzero;
+    cos_a.h = onepointzero;
+    cos_a.l = zeropointzero;
+    return;
+  }
+
+  sin_a = sin_taylor(input);
+  cos_a = sqrt(sub(onepointzero, sqr(sin_a)));
+}
+
+// Reference: QD / dd_real.cpp
+template <typename T>
+inline two<T> sin(const two<T> &input) {
+
+  two<T> local_2pi;
+  two<T> local_pi2;
+  two<T> local_pi16;
+  const T* local_ptr_cos_table;
+  const T* local_ptr_sin_table;
+  T zeropointzero, pointfive, local_nan;
+
+  if constexpr (std::is_same_v<T, float>) {
+    local_2pi = fp32_2pi;
+    local_pi2 = fp32_pi2;
+    local_pi16 = fp32_pi16;
+    local_ptr_cos_table = &fp32_cos_table[0][0];
+    local_ptr_sin_table = &fp32_sin_table[0][0];
+    zeropointzero = 0.0f;
+    pointfive  = 0.5f;
+    local_nan  = fp32_nan;
+  } else if constexpr (std::is_same_v<T, double>) {
+    local_2pi = fp64_2pi;
+    local_pi2 = fp64_pi2;
+    local_pi16 = fp64_pi16;
+    local_ptr_cos_table = &fp64_cos_table[0][0];
+    local_ptr_sin_table = &fp64_sin_table[0][0];
+    zeropointzero = 0.0;
+    pointfive  = 0.5;
+    local_nan  = fp64_nan;
+  } else {
+    std::cerr << "Other types not supported\n";
+  }
+
+  if (input.eval() == zeropointzero) {
+    T hi, lo;
+    qd::split(zeropointzero, hi, lo);
+    return two<T>{hi, lo};
+  }
+
+  // Approximately reducing modulo 2*pi
+  two<T> z = nint(div<doubleword::Mode::Accurate, true>(input, local_2pi)); // TODO: check mode chosen for div (accurate + fma)
```

Comment:
The user needs to be able to specify via a template argument whether he wants to use the fma version or not, as FMA instructions are not available on some platforms (eg IPUs)

Reply:
The `sin()` function has been parameterized for mode and useFMA in commit [912281b](https://github.com/esa-tu-darmstadt/twofloat/pull/1/commits/912281b3fe321362efa66dc6980b1ed80ae1d588).